lisa/tests/base.py

```diff
@@ -989,8 +990,22 @@ def trace_window(self, trace):
           Calling ``self.trace`` here will raise an :exc:`AttributeError`
           exception, to avoid entering infinite recursion.
         """
-        # Find when the first rtapp phase starts
-        rta_start = trace.analysis.rta.df_rtapp_phases_start()['Time'].min()
+        swdf = trace.df_events('sched_switch')
+
+        def get_first_switch(swdf, idx, row):
+            comm, pid, _ = idx
+            start_time = row['Time']
+            task = TaskID(comm=comm, pid=pid)
+            start_swdf = df_filter_task_ids(swdf, [task], pid_col='next_pid', comm_col='next_comm')
+            return start_swdf[start_swdf.index < start_time].index[-1]
+
+        # Find when the first rtapp phase starts, and take the associated
+        # sched_switch that is immediately preceding
+        rta_start = min(
+            get_first_switch(swdf, idx, row)
+            for idx, row in trace.analysis.rta.df_rtapp_phases_start().iterrows()
```

Comment:
`iterrows()` should be avoided as much as possible - it's a performance killer (and isn't even guaranteed to yield every row exactly once...). Now those are rtapp phases so we shouldn't have a gazillion of them, but perhaps this would be slightly better (will require tuning the inline function parameters):

```
df_phase_switch =  trace.analysis.rta.df_rtapp_phases_start().apply(get_first_switch())
rta_start = df_phase_switch.min
```

Maybe you can `timeit.timeit` and see if it's really noticeable.

Reply:
`timeit.timeit` wont show anything since my df has less than 10 rows, but I'll make the change if that works (also simplifies the code)